drivers: move disk drivers (and sdmmc controller drivers) to drivers/disk

jfischer-no opened the pull request:
The files disk_access_usdhc.c, disk_access_spi_sdhc.c, and
disk_access_stm32_sdmmc.c are actually drivers for SD/MMC
controllers. This patch moves this drivers to drivers/sdmmc and
reworks the configuration so that the drivers are selected when
the corresponding node is enabled.

jfischer-no committed:
samples: mass: add SD card support

Add SD card support to USB MSC sample.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
tests/boards/samples: fixup after sdmmc driver relocation

Fixup configuration.
Remove obsolete SDHC SPI configuration.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
disk: move disk and sdmmc controller drivers to drivers/disk

The files disk_access_usdhc.c, disk_access_spi_sdhc.c,
disk_access_stm32_sdmmc.c, disk_access_ram.c and
disk_access_flash.c are actually drivers for block devices and SD/MMC
controllers. This patch moves this drivers to drivers/disk and
reworks the configuration so that the drivers are selected when
the corresponding node is enabled.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: disk: cleanup after driver relocation

Add common SDMMC_LOG_LEVEL and SDMMC_VOLUME_NAME.
Initialize drivers at POST_KERNEL level.
Update CODEOWNERS after sdmmc drivers relocation.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: usdhc: fixup i.MX RT related code after driver relocation

Fixup i.MX RT related code after driver relocation.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

erwango commented:
^^ @anthonybrandon

jfischer-no committed:
CODEOWNERS: add myself as disk subsystem code-owner

Add myself as disk subsystem code-owner

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

anthonybrandon commented:
I haven't tested on a board yet, but so far it looks fine.

galak commented:
Dev-review (Feb 25): 
* split out driver related bits in include/disk/disk_access.h into include/drivers/disk.h
* Move drivers out of subsys/disk/{flash,ram,*sd*, *mmc*} into drivers/disk

jfischer-no committed:
include: move disk driver interface to own header

Move disk driver interface to own header and
separate disk access interface and disk driver interface.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
include: move disk_access.h to storage

Zephyr already has a directory for storage API relevant headers.
Move disk_access.h header to include/storage where it fits better
structurally.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
doc: update disk API documentation

Fix disk access interface description and
update disk API documentation.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
doc: update disk access docu regarding SD card support

This patch removes sdhc_api references since such API
does not actually exist. SD card controller drivers
use disc driver interface. Accordingly, the SDHC documentation,
which actually describes how to connect a SD card via SPI
and has little to do with SD host controller (SDHC),
is also moved to disk access page and adapted.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

carlescufi reviewed (approved):
OK on the move/rename as agreed during the dev-review meeting.

carlescufi commented:
@MaureenHelm could you please review?

MaureenHelm commented on boards/arm/mimxrt1050_evk/pinmux.c:
```diff
@@ -17,7 +17,7 @@ static gpio_pin_config_t enet_gpio_config = {
 #endif
 
-#if DT_NODE_HAS_STATUS(DT_NODELABEL(usdhc1), okay) && CONFIG_DISK_ACCESS_USDHC1
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(usdhc1), okay) && CONFIG_DISK_DRIVER_SDMMC
```
No

MaureenHelm reviewed (changes requested):
Looks good. Tested `samples/subsys/usb/mass` and `samples/subsys/fs/fat_fs` on `mimxrt1064_evk`.

Please also update the release notes

jfischer-no commented on drivers/disk/Kconfig.sdmmc:
```diff
@@ -0,0 +1,59 @@
+
+config DISK_DRIVER_SDMMC
+	bool "SDMMC card driver"
+	select FLASH
```
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/disk/Kconfig#L97, but that is good question. These depends and selects around FLASH* overall in the tree annoy me a lot.

jfischer-no replied:
fixed

jfischer-no committed:
doc: release-notes-2.6: add note about disk drivers rework

Add note about disk drivers rework.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>
